File: mobile/src/test/java/com/alexstyl/specialdates/upcoming/UpcomingRowViewModelsBuilderTest.java
package com.alexstyl.specialdates.upcoming;

import com.alexstyl.resources.ColorResources;
import com.alexstyl.resources.StringResources;
import com.alexstyl.specialdates.DisplayName;
import com.alexstyl.specialdates.Optional;
import com.alexstyl.specialdates.contact.DeviceContact;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.date.DateDisplayStringCreator;
import com.alexstyl.specialdates.date.TimePeriod;
import com.alexstyl.specialdates.events.bankholidays.BankHoliday;
import com.alexstyl.specialdates.events.namedays.NamesInADate;
import com.alexstyl.specialdates.events.peopleevents.StandardEventType;

import java.util.Arrays;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static com.alexstyl.specialdates.date.DateConstants.*;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

@RunWith(MockitoJUnitRunner.class)
public class UpcomingRowViewModelsBuilderTest {

    private static final Date FEBRUARY_1st = Date.on(1, FEBRUARY, 1990);
    private static final Date FEBRUARY_3rd = Date.on(3, FEBRUARY, 1990);
    private static final Date MARCH_5th = Date.on(5, MARCH, 1990);
    private static final Optional<Long> NO_DEVICE_EVENT_ID = Optional.absent();

    private UpcomingRowViewModelsFactory upcomingRowViewModelsFactory;

    @Mock
    private ColorResources mockColorResources;
    @Mock
    private StringResources mockStringResources;

    @Before
    public void setUp() {
        Date today = Date.today();
        upcomingRowViewModelsFactory = new UpcomingRowViewModelsFactory(
                today,
                DateDisplayStringCreator.INSTANCE,
                new ContactViewModelFactory(mockColorResources, mockStringResources),
                mockStringResources,
                new BankHolidayViewModelFactory(),
                new NamedaysViewModelFactory(today)
        );

    }

    @Test
    public void celebrationDateIsCreatedCorrectlyForDifferentYear() {
        ContactEvent event = aContactEventOn(Date.on(1, JANUARY, 1990));
        TimePeriod duration = TimePeriod.between(Date.on(1, JANUARY, 2016), Date.on(1, DECEMBER, 2016));

        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(duration, upcomingRowViewModelsFactory)
                .withContactEvents(singletonList(event))
                .build();

        assertThat(dates.size()).isEqualTo(1);
    }

    @Test
    public void givenASingleContactEvent_thenOneCelebrationDateIsCreated() {
        ContactEvent event = aContactEvent();
        List<ContactEvent> contactEvents = singletonList(event);

        TimePeriod duration = timeOf(event);

        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(duration, upcomingRowViewModelsFactory)
                .withContactEvents(contactEvents)
                .build();

        assertThat(dates.size()).isEqualTo(1);
    }

    @Test
    public void givenTwoContactEventsOnSameDate_thenOneCelebrationDateIsCreated() {
        List<ContactEvent> contactEventsList = Arrays.asList(
                aContactEventOn(FEBRUARY_1st),
                aContactEventOn(FEBRUARY_1st)
        );

        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(TimePeriod.between(FEBRUARY_1st, FEBRUARY_1st), upcomingRowViewModelsFactory)
                .withContactEvents(contactEventsList)
                .build();

        assertThat(dates.size()).isEqualTo(1);
    }

    @Test
    public void givenTwoContactEventsOnDifferentDate_thenTwoCelebrationDateAreCreated() {
        List<ContactEvent> contactEventsList = Arrays.asList(
                aContactEventOn(FEBRUARY_1st),
                aContactEventOn(FEBRUARY_3rd)
        );

        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(TimePeriod.between(FEBRUARY_1st, FEBRUARY_3rd), upcomingRowViewModelsFactory)
                .withContactEvents(contactEventsList)
                .build();

        assertThat(dates.size()).isEqualTo(2);
    }

    @Test
    public void givenABankHoliday_thenACelebrationDateIsCreated() {
        BankHoliday bankHoliday = aBankHoliday();
        List<BankHoliday> bankHolidays = singletonList(bankHoliday);
        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(TimePeriod.between(bankHoliday.getDate(), bankHoliday.getDate()), upcomingRowViewModelsFactory)
                .withBankHolidays(bankHolidays)
                .build();

        assertThat(dates.size()).isEqualTo(1);
    }

    @Test
    public void givenEventsOnDifferentEvents_thenACelebrationDatesForEachOneAreCreated() {
        List<UpcomingRowViewModel> dates = new UpcomingRowViewModelsBuilder(TimePeriod.between(FEBRUARY_1st, MARCH_5th), upcomingRowViewModelsFactory)
                .withContactEvents(singletonList(aContactEventOn(FEBRUARY_1st)))
                .withBankHolidays(singletonList(aBankHolidayOn(FEBRUARY_3rd)))
                .withNamedays(singletonList(new NamesInADate(MARCH_5th, singletonList("Name"))))
                .build();

        assertThat(dates.size()).isEqualTo(3);
    }

    private static BankHoliday aBankHolidayOn(Date date) {
        return new BankHoliday("A bank holiday", date);
    }

    private static BankHoliday aBankHoliday() {
        return new BankHoliday("A bank holiday", Date.on(1, JANUARY, 1990));
    }

    private static ContactEvent aContactEvent() {
        return aContactEventOn(Date.on(1, JANUARY, 1990));
    }

    private static ContactEvent aContactEventOn(Date date) {
        DeviceContact contact = new DeviceContact(1, DisplayName.NO_NAME, "lookup_key");
        return new ContactEvent(NO_DEVICE_EVENT_ID, StandardEventType.BIRTHDAY, date, contact);
    }

    private static TimePeriod timeOf(ContactEvent event) {
        return TimePeriod.between(event.getDate(), event.getDate());
    }

}

File: memento/src/test/java/com/alexstyl/specialdates/upcoming/UpcomingRowViewModelsBuilderTest.kt
package com.alexstyl.specialdates.upcoming

import com.alexstyl.TestColors
import com.alexstyl.resources.Colors
import com.alexstyl.specialdates.JavaStrings
import com.alexstyl.specialdates.contact.ContactFixture.aContact
import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.date.Months
import com.alexstyl.specialdates.date.Months.FEBRUARY
import com.alexstyl.specialdates.date.Months.JANUARY
import com.alexstyl.specialdates.date.TimePeriod
import com.alexstyl.specialdates.date.beggingOfYear
import com.alexstyl.specialdates.date.dateOn
import com.alexstyl.specialdates.date.endOfYear
import com.alexstyl.specialdates.date.todaysDate
import com.alexstyl.specialdates.events.bankholidays.BankHoliday
import com.alexstyl.specialdates.events.namedays.ArrayNamesInADate
import com.alexstyl.specialdates.events.namedays.NamesInADate
import com.alexstyl.specialdates.events.peopleevents.StandardEventType
import org.fest.assertions.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.BDDMockito.given
import org.mockito.Mock
import org.mockito.runners.MockitoJUnitRunner
import java.util.Arrays.asList

@RunWith(MockitoJUnitRunner::class)
class UpcomingRowViewModelsBuilderTest {

    private var upcomingEventRowViewModelFactory: UpcomingEventRowViewModelFactory? = null
    private val mockStrings = JavaStrings()

    @Mock
    private val mockColors: Colors? = TestColors()
    private val mockDateStringCreator = DateStringCreator()

    @Before
    fun setUp() {
        given(mockColors!!.getTodayHeaderTextColor()).willReturn(5)
        val today = todaysDate()
        upcomingEventRowViewModelFactory = UpcomingEventRowViewModelFactory(
                today,
                mockColors,
                mockDateStringCreator,
                ContactViewModelFactory(mockColors, mockStrings)
        )

    }

    @Test
    fun givenASingleContactEvent_thenCreatesADateHeaderPlusAContactEvent() {

        val viewModels = builderFor(entireYear(2016))
                .withContactEvents(asList(aContactEventOn(dateOn(1, JANUARY, 2016))))
                .build()

        assertThat(viewModels.size).isEqualTo(2)
        assertThat(viewModels[0]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[1]).isInstanceOf(UpcomingContactEventViewModel::class.java)
    }

    @Test
    fun givenTwoContactEvents_whenBothAreOnTheSameDate_thenCreatesADateHeaderPlusTwoContactEvents() {
        val viewModels = builderFor(entireYear(1990))
                .withContactEvents(asList(
                        aContactEventOn(dateOn(1, FEBRUARY, 1990)),
                        aContactEventOn(dateOn(1, FEBRUARY, 1990))
                ))
                .build()

        assertThat(viewModels.size).isEqualTo(3)
        assertThat(viewModels[0]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[1]).isInstanceOf(UpcomingContactEventViewModel::class.java)
        assertThat(viewModels[2]).isInstanceOf(UpcomingContactEventViewModel::class.java)
    }

    @Test
    fun givenTwoContactEvents_whenBothAreOnDifferentDate_thenCreatesADateHeaderAndAContactEventForEachOne() {
        val viewModels = builderFor(entireYear(1990))
                .withContactEvents(asList(
                        aContactEventOn(dateOn(1, FEBRUARY, 1990)),
                        aContactEventOn(dateOn(3, FEBRUARY, 1990))
                ))
                .build()

        assertThat(viewModels.size).isEqualTo(4)
        assertThat(viewModels[0]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[1]).isInstanceOf(UpcomingContactEventViewModel::class.java)
        assertThat(viewModels[2]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[3]).isInstanceOf(UpcomingContactEventViewModel::class.java)
    }

    @Test
    fun givenABankHoliday_thenCreatesADateHeaderPlusABankholidayModel() {
        val viewModels = builderFor(TimePeriod.between(aBankHoliday().date, aBankHoliday().date))
                .withBankHolidays(asList(aBankHoliday()))
                .build()

        assertThat(viewModels.size).isEqualTo(2)
        assertThat(viewModels[0]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[1]).isInstanceOf(BankHolidayViewModel::class.java)
    }

    @Test
    fun givenANameday_thenCreatesADateHeaderPlusANamedayModel() {
        val date = dateOn(1, Months.APRIL, 2017)
        val namedays = namedayOf(date, "Maria")

        val viewModels = builderFor(TimePeriod.between(date, date))
                .withNamedays(namedays)
                .build()

        assertThat(viewModels.size).isEqualTo(2)
        assertThat(viewModels[0]).isInstanceOf(DateHeaderViewModel::class.java)
        assertThat(viewModels[1]).isInstanceOf(UpcomingNamedaysViewModel::class.java)
    }

    @Test
    fun theDisplayOrderIsCorrect() {
        val date = dateOn(1, Months.APRIL, 2017)

        val bankHoliday = BankHoliday("A bank holiday", date)
        val bankHolidays = listOf(bankHoliday)
        val namedays = namedayOf(date, "Maria")

        val viewModels = builderWithAds(entireYear(2017))
                .withContactEvents(asList(
                        aContactEventOn(dateOn(1, Months.APRIL, 2017)),
                        aContactEventOn(dateOn(2, Months.APRIL, 2017))
                ))
                .withNamedays(namedays)
                .withBankHolidays(bankHolidays)
                .build()

        assertListContainsOrderedItems(viewModels, asList(
                DateHeaderViewModel::class.java,
                BankHolidayViewModel::class.java,
                UpcomingNamedaysViewModel::class.java,
                UpcomingContactEventViewModel::class.java,
                DateHeaderViewModel::class.java,
                UpcomingContactEventViewModel::class.java
        ))
    }

    private fun assertListContainsOrderedItems(viewModels: List<UpcomingRowViewModel>,
                                               classes: List<Class<out UpcomingRowViewModel>>) {
        classes.forEachIndexed { index, clazz ->
            assertThat(viewModels[index]).isInstanceOf(clazz)
        }
        assertThat(viewModels.size).isEqualTo(classes.size)
    }

    private fun namedayOf(date: Date, maria: String): List<NamesInADate> {
        return listOf(ArrayNamesInADate(date, arrayListOf(maria)))
    }

    private fun builderFor(timePeriod: TimePeriod): UpcomingRowViewModelsBuilder {
        return UpcomingRowViewModelsBuilder(timePeriod, upcomingEventRowViewModelFactory!!)
    }

    private fun builderWithAds(timePeriod: TimePeriod): UpcomingRowViewModelsBuilder {
        return UpcomingRowViewModelsBuilder(timePeriod, upcomingEventRowViewModelFactory!!)
    }


    private fun aBankHoliday(): BankHoliday {
        return BankHoliday("A bank holiday", dateOn(1, JANUARY, 1990))
    }

    private fun aContactEventOn(date: Date): ContactEvent {
        return ContactEvent(StandardEventType.BIRTHDAY, date, aContact(), NO_DEVICE_EVENT_ID)
    }

    private fun entireYear(year: Int): TimePeriod {
        return TimePeriod.between(beggingOfYear(year), endOfYear(year))
    }

    companion object {
        private val NO_DEVICE_EVENT_ID: Long? = null
    }

}


class DateStringCreator : UpcomingDateStringCreator {
    override fun createLabelFor(date: Date): String = "Header Label"

}
